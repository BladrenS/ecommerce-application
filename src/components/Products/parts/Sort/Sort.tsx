import clsx from 'clsx';
import { type FC } from 'react';

import { useCatalogContext } from '../../../../pages/Catalog';
import { type SortValue } from '../../../../pages/Catalog/types';
import styles from './styles.module.scss';

type SortValueType = SortValue['value'];

const SORTING_VALUE: SortValueType[] = ['default', 'price', 'name'];

export const Sort: FC = () => {
  const {
    sort: { value, direction },
    setSort,
  } = useCatalogContext();

  const clickSortHandler = (clickedValue: SortValueType) => {
    if (clickedValue === value && clickedValue === 'default') return;

    setSort((previous) => {
      const isSameField = previous.value === clickedValue;

      const updated: SortValue = {
        value: clickedValue,
        direction: isSameField ? (previous.direction === 'asc' ? 'desc' : 'asc') : 'asc',
      };

      return updated;
    });
  };

  return (
    <div className={styles.container}>
      <div className={styles.title}>Sorting:</div>
      <div className={styles.wrapper}>
        {SORTING_VALUE.map((currentValue) => (
          <div className={clsx(styles.value, {})} onClick={() => clickSortHandler(currentValue)} key={currentValue}>
            <div
              className={clsx(styles.item, {
                [styles.active]: value === currentValue,
              })}
            >
              {currentValue}
            </div>
            <span className={styles.decor}>
              {currentValue === value && currentValue !== 'default' ? (direction === 'asc' ? 'ðŸ •' : 'ðŸ —') : ''}
            </span>
          </div>
        ))}
      </div>
    </div>
  );
};
