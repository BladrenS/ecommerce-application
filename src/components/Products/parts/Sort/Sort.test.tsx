import { render, screen } from '@testing-library/react';

import { Sort } from './Sort';

jest.mock('../../../../pages/Catalog', () => ({
  useCatalogContext: () => ({
    filters: {
      sort: { value: 'default', direction: 'asc' },
    },
    setFilters: jest.fn(),
  }),
}));

describe('Sort snapshot', () => {
  it('соответствует снапшоту', () => {
    const { asFragment } = render(<Sort />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Sort', () => {
  test('рендерит все варианты сортировки', () => {
    render(<Sort />);
    expect(screen.getByText('Sorting:')).toBeInTheDocument();
    expect(screen.getByText('default')).toBeInTheDocument();
    expect(screen.getByText('price')).toBeInTheDocument();
    expect(screen.getByText('name')).toBeInTheDocument();
  });
});
