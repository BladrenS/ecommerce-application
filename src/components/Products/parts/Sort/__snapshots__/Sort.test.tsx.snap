// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort snapshot соответствует снапшоту 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="title"
    >
      Sorting:
    </div>
    <div
      class="wrapper"
    >
      <div
        class="value"
      >
        <div
          class="item active"
        >
          default
        </div>
        <span
          class="decor"
        />
      </div>
      <div
        class="value"
      >
        <div
          class="item"
        >
          price
        </div>
        <span
          class="decor"
        />
      </div>
      <div
        class="value"
      >
        <div
          class="item"
        >
          name
        </div>
        <span
          class="decor"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
